src/tqec/templates/scale.py

```diff
@@ -12,110 +17,347 @@ class LinearFunction:
     def __call__(self, x: int) -> int:
         return self.slope * x + self.offset
 
-    def to_dict(self) -> dict[str, ty.Any]:
-        return {"type": type(self).__name__, "slope": self.slope, "offset": self.offset}
-
-    def __add__(self, other: "LinearFunction") -> "LinearFunction":
+    def __add__(self, other: LinearFunction | int) -> LinearFunction:
+        if isinstance(other, int):
+            other = LinearFunction(0, other)
         return LinearFunction(self.slope + other.slope, self.offset + other.offset)
 
-    def __sub__(self, other: "LinearFunction") -> "LinearFunction":
+    def __sub__(self, other: LinearFunction | int) -> LinearFunction:
+        if isinstance(other, int):
+            other = LinearFunction(0, other)
         return LinearFunction(self.slope - other.slope, self.offset - other.offset)
 
-    def __mul__(self, other: int) -> "LinearFunction":
+    def __mul__(self, other: int) -> LinearFunction:
         return self.__rmul__(other)
 
-    def __rmul__(self, other: int) -> "LinearFunction":
+    def __rmul__(self, other: int) -> LinearFunction:
         return LinearFunction(other * self.slope, other * self.offset)
 
-    def invert(self, value: int) -> int:
-        if self.slope == 0:
-            raise TQECException(
-                f"Cannot invert {self}: the linear function is constant and so "
-                f"has an infinite set of inverse for {self.offset} and no inverse "
-                "for any other value."
-            )
-        offset_value = value - self.offset
-        if offset_value % self.slope != 0:
-            raise TQECException(
-                f"{self} cannot invert exactly {value} as this would lead to "
-                f"a non-integer value ({offset_value / self.slope})."
-            )
-        return (value - self.offset) // self.slope
+    def intersection(self, other: LinearFunction) -> float | None:
+        """Compute the intersection between two linear functions.
+        Args:
+            other: the `LinearFunction` instance to intersect with `self`.
+        Returns:
+            an instance of `IntersectionResults` containing the necessary
+            data to know when the two `LinearFunction` instances intersect
+            and which one is above/below before/after the intersection.
```

Comment:
```suggestion
            If they intersect, return x such that `self(x) = other(x)`.
            Otherwise, return None.
```
`IntersectionResults` is not defined.